<launch>
	<!-- Monitors -->
	<node name="rosbag_recorder" pkg="rosbag" type="record" args="-a -e fmData|usageMonitor -O $(find sdu_pichi_remote)/bags/test.bag" />
	<node name="usageMonitor" pkg="usage_monitor" type="usage_monitor_node">
		<param name="rate" value="4" />
  	</node>
  	
	<!-- Transforms -->
	<node pkg="tf" type="static_transform_publisher" name="map_broadcaster" args="-588599.0922487933 -6137322.245287642 0 0 0 0 gps_frame map 100"/>
	<node pkg="tf" type="static_transform_publisher" name="mast_broadcaster" args="-1.20 0 0 0 0 0 mast_top mast_bottom 100"/>	
	<node pkg="sdu_pichi_remote" type="imu_tf_publisher.py" name="imu_broadcaster"/>
	

	<!-- Controllers -->
	<group ns="fmControllers">
		<node pkg="differential_ifk" name="diff" type="differential_ifk_node" output="screen" >
			<param name="hl_subscriber_topic" value="/fmSignals/cmd_vel"/>
			<param name="ll_publisher_topic_left" value="/fmSignals/cmd_vel_left"/>
			<param name="ll_publisher_topic_right" value="/fmSignals/cmd_vel_right"/>
			<param name="distance_center_to_wheel" value="0.42" />
		</node>
	</group>

	<!-- Platform -->
	<group ns="robot">
		<node pkg="sdu_pichi" name="sdu_pichi_node" type="sdu_pichi_node" output="screen" >
			<param name="cmd_vel_ch1_topic" value="/fmSignals/cmd_vel_left"/>
			<param name="cmd_vel_ch2_topic" value="/fmSignals/cmd_vel_right"/>
			<param name="serial_rx_topic" value="/fmData/rx"/>
			<param name="serial_tx_topic" value="/fmData/tx"/>
			<param name="deadman_topic" value="/fmSignals/deadman"/>
			<param name="encoder_ch1_topic" value="/fmInformation/encoder_left"/>
			<param name="encoder_ch2_topic" value="/fmInformation/encoder_right"/>	
			<param name="power_ch1_topic" value="/fmData/power_left"/>
			<param name="power_ch2_topic" value="/fmData/power_right"/>
			<param name="status_topic" value="/fmData/status"/>	
			<param name="mps_to_rpm" value="1432"/>
			<param name="closed_loop_operation" value="true"/>
			<param name="max_acceleration" value="20000"/>
			<param name="max_deceleration" value="20000"/>
			<param name="max_rpm" value="4000"/>
			<param name="anti_windup_percent" value="50"/>
			<param name="p_gain" value="0"/>
			<param name="i_gain" value="0"/>
			<param name="d_gain" value="0"/>         
        </node>
		
	    <node pkg="serial_string" name="serial_string_node" type="serial_string_node" output="screen" respawn="true" >
	        <param name="publisher_topic" value="/fmData/pichi_rx" />	 
	        <param name="subscriber_topic" value="/fmData/pichi_tx" />	
	        <param name="device" value="/dev/pichi" />	
	        <param name="baudrate" value="115200" />
	        <param name="termination_character" value="13"/>
		</node>	
	</group>
	
	<!-- Sensors -->
	<group ns="platform_feedback">
		<!-- GPS -->
		<node pkg="serial_string" name="gps_serial" type="serial_string_node" output="screen"  respawn="true">
			<param name="publisher_topic" value="/fmData/gps_rx" />	 
			<param name="subscriber_topic" value="/fmData/gps_tx" />
			<param name="device" value="/dev/gps" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node>

		<!-- IMU -->
		<node pkg="serial_string" name="imu_serial" type="serial_string_node" output="screen" respawn="true">
			<param name="publisher_topic" value="/fmData/imu_rx" /> 
			<param name="subscriber_topic" value="/fmData/imu_tx" />
			<param name="device" value="/dev/imu" />	
			<param name="baudrate" value="115200" />
			<param name="termination_character" value="10"/>
		</node> 
		
		<node pkg="vectornav_vn100" name="imu_node"  type="vectornav_vn100_node" output="screen" respawn="True" >
			<param name="vectornav_vn100_sub" value="/fmData/imu_rx"/>
			<param name="imu_pub" value="/fmInformation/imu"/>
			<param name="frame_id" value="mast_top"/>
			<param name="use_enu" value="true"/>
			<param name="covariance_x" value="0.001"/>
			<param name="covariance_y" value="0.001"/>
			<param name="covariance_z" value="0.001"/>
		 </node>	
	</group>
	
	<!-- Processors -->
	<group ns="platform_processing">
		<!-- Vehicle odometry -->
		<!--node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/enc_left" />
			<param name="enc_right_sub" value="/fmInformation/enc_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_pub" value="/fmKnowledge/encoder_odom" />
			<param name="yaw_angle_source" value="imu_angular_velocity"/>  
			<param name="imu_angular_velocity_yaw_axis" value="-d" /> 
			<param name="vehicle_frame" value="base_footprint" />      
			<param name="odom_estimate_frame" value="/wheel_odom" />      
		</node-->
		
		<!-- GPS odometry -->
		<node pkg="gps_gpgga" name="nmea_to_gpgga" type="gps_gpgga_node" output="screen">
        	<param name="subscribe_topic_id" value="/fmData/gps_rx" />
			<param name="publish_topic_id" value="/fmInformation/gpgga" />
		</node>
	
    	<node pkg="gpgga_to_utm" name="gpgga_to_utm_converter" type="gpgga_to_utm_node" output="screen"> 
        	<param name="subscribe_topic_id" value="/fmInformation/gpgga" />
			<param name="publish_topic_id" value="/fmInformation/utm" />
		</node>
	
    	<node pkg="utm_to_odom" name="gps_odometry" type="utm_to_odom_node" output="screen"> 
    		<param name="subscribe_topic_id" value="/fmInformation/utm" />
			<param name="publish_topic_id" value="/fmKnowledge/gps_odom" />
			<param name="odom_frame_id" value="gps_frame" />
			<param name="vehicle_frame" value="gps_frame" />
			<param name="odom_estimate_frame" value="world_frame" />
			<param name="child_odom_frame_id" value="world_frame" />
			<param name="receive_n_before_publish" value="10"/>
			<param name="gps_variance" value="1"/>
		</node>
	</group>
</launch>
